src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,129 @@
+import React, { memo, useState } from 'react';
+import cN from 'classnames';
+import { Todo } from '../../types/Todo';
+import { TodoPatch } from '../../types/TodoPatch';
+
+type Props = {
+  todo: Todo,
+  onDelete: (todoId: number) => void,
+  onUpdateTodo?: (todoId: number, data: TodoPatch) => void,
+  isLoading?: boolean
+};
+
+export const TodoItem: React.FC<Props> = memo(
+  ({
+    todo,
+    onDelete,
+    onUpdateTodo = () => { },
+    isLoading = true,
+  }) => {
+    const {
+      id,
+      title,
+      completed,
+    } = todo;
+
+    const [isEditing, setIsEditing] = useState(false);
+    const [newTitle, setNewTitle] = useState(title);
+
+    const handleDeleteTodo = () => onDelete(id);
+
+    const handleToggleTodo = () => onUpdateTodo(id, { completed: !completed });
+
+    const handleToggleEgiting = () => setIsEditing(state => !state);
+
+    const handleChangeNewTitle = (
+      event: React.ChangeEvent<HTMLInputElement>,
+    ) => {
+      setNewTitle(event.target.value);
+    };
+
+    const handleSubmit = (event: React.FormEvent) => {
+      event.preventDefault();
+
+      if (!newTitle.trim().length) {
+        onDelete(id);
+      } else if (title !== newTitle) {
+        setNewTitle(prevTitle => prevTitle.trim());
+        onUpdateTodo(id, { title: newTitle });
+      }
+    };
+
+    const handleBlur = (event: React.FormEvent) => {
+      handleToggleEgiting();
+      handleSubmit(event);
+    };
+
+    const handleCancelEditing = (
+      event: React.KeyboardEvent<HTMLInputElement>,
+    ) => {
+      if (event.key === 'Escape') {
+        setIsEditing(false);
+        setNewTitle(title);
+      }
+    };
+
+    return (
+      <div
+        className={cN('todo', {
+          completed,
+        })}
+        key={todo.id}
```

Comment:
`Key` prop is unnecessary in this component as it is not part of an array or iterator.